packages/astro/src/core/request.ts

```diff
@@ -0,0 +1,32 @@
+import type { IncomingHttpHeaders } from 'http';
+
+type HeaderType = Headers | Record<string, any> | IncomingHttpHeaders;
+
+export function createRequest(url: URL | string, headers: HeaderType, method = 'GET'): Request {
+	let headersObj = headers instanceof Headers ? headers :
+		new Headers(Object.entries(headers as Record<string, any>));
+	
+	const request = new Request(url.toString(), {
+		method: method,
+		headers: headersObj
+	});
+
+	Object.defineProperties(request, {
+		canonicalURL: {
+			get() {
+				console.warn(`Astro.request.canonicalURL has been moved to Astro.canonicalURL`);
```

Comment:
Nit: should we be using `logger` instead here? Not certain

Reply:
Yeah I guess we should, I'm not sure why I did it this way. Can change.